pkg/server/jwk_resolver.go

```diff
@@ -0,0 +1,129 @@
+// Copyright 2023 The Authors (see AUTHORS file)
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//	http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+package server
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	"net/http"
+	"net/url"
+	"slices"
+	"sync"
+	"time"
+
+	"github.com/lestrrat-go/jwx/v2/jwk"
+	"github.com/lestrrat-go/jwx/v2/jwt"
+
+	"github.com/abcxyz/pkg/cache"
+)
+
+type JWKResolver interface {
+	ResolveKeySet(ctx context.Context, oidcHeader string) (jwk.Set, error)
+}
+
+type OIDCResolver struct {
+	issuerAllowlist []string
+	cache           *jwk.Cache
+	issuerToJwksURI *cache.Cache[string]
+	mu              sync.Mutex
+}
+
+type OpenIDConfiguration struct {
+	// The only field we need from the config
+	JwksURI string `json:"jwks_uri"`
+}
+
+func NewOIDCResolver(ctx context.Context, issuerAllowlist []string, jwksURICacheTimeout time.Duration) *OIDCResolver {
+	return &OIDCResolver{
+		issuerAllowlist: issuerAllowlist,
+		cache:           jwk.NewCache(ctx),
+		issuerToJwksURI: cache.New[string](jwksURICacheTimeout),
+	}
+}
+
+// ResolveKeySet finds and caches the correct OIDC key set to use based on the token's issuer's OpenID Configuration.
+func (r *OIDCResolver) ResolveKeySet(ctx context.Context, oidcHeader string) (jwk.Set, error) {
+	// Parse without verification to extract issuer so we can use it to obtain the key set to use for verification
+	token, err := jwt.ParseString(oidcHeader, jwt.WithContext(ctx), jwt.WithVerify(false), jwt.WithValidate(false))
+	if err != nil {
+		return nil, fmt.Errorf("failed to parse issuer from jwt header: %w", err)
+	}
+	issuer := token.Issuer()
+
+	if !slices.Contains(r.issuerAllowlist, issuer) {
```

Comment:
Used a string slice, mainly because there's no flag convenience function for `map[string]struct{}` and `map[string]string` seems counterintuitive to me. Do you think we should go and add a convenience function for `StringSetVar` that resolves to `map[string]struct{}`? I can do that if you think it would be better.

Reply:
It's not worth the optimization, especially given N is very small.